core/lib/types/src/event.rs

```diff
@@ -339,6 +344,73 @@ pub struct VmEventGroupKey {
     pub topic: (u32, H256),
 }
 
+/// Each `VmEvent` can be translated to several log queries.
+/// This methods converts each event from input to log queries and returns all produced log queries.
+pub fn convert_vm_events_to_log_queries(events: &[VmEvent]) -> Vec<LogQuery> {
+    events
+        .iter()
+        .flat_map(|event| {
+            // Construct first query. This query holds an information about
+            // - number of event topics (on log query level `event.address` is treated as a topic, thus + 1 is added)
+            // - length of event value
+            // - `event.address` (or first topic in terms of log query terminology).
+            let first_key_word =
+                (event.indexed_topics.len() as u64 + 1) + ((event.value.len() as u64) << 32);
+            let key = U256([first_key_word, 0, 0, 0]);
+            let first_log = LogQuery {
+                timestamp: Timestamp(0),
+                tx_number_in_block: event.location.1 as u16,
+                aux_byte: 0,
+                shard_id: 0,
+                address: EVENT_WRITER_ADDRESS,
+                key,
+                read_value: U256::zero(),
+                written_value: address_to_u256(&event.address),
+                rw_flag: false,
+                rollback: false,
+                is_service: true,
+            };
+
+            // The next logs hold an information about remaining topics and `event.value`.
+            // Each log can hold at most two values each of 32 bytes.
+            // The following piece of code prepares these 32-byte values.
+            let values = event
+                .indexed_topics
+                .iter()
+                .map(|h| h256_to_u256(*h))
+                .chain(event.value.chunks(32).map(U256::from_big_endian));
```

Comment:
I am not 100% sure `U256::from_big_endian` will work fine with `value` that is not divisible by 32

Reply:
Is it possible that value is not divisible by 32? I couldn't find an example that gives `value.len() % 32 != 0`

Reply:
Maybe noone does that because a typical Solidity contract uses ABI encoded values which pad bytes with 0s to ensure that they are divisible by 32. However, there is nothing that in principle fobids value from being 1 byte long